contrib/bindings/go/pathrs/pathrs.go

```diff
@@ -0,0 +1,294 @@
+	h.fileCounter++
+	return name
+}
+
+// Error representation of rust error
+// particularly useful to not frighten to lost controll of pointer which can be rewritten.
+type Error struct {
+	description string
+	errno       uint64
+	backtrace   []backtraceLine
+}
+
+type backtraceLine struct {
+	ip       uintptr
+	sAddress uintptr
+	sName    string
+	sFile    string
+	sLineno  uint32
+}
+
+func (err *Error) Error() string {
+	return err.description
+}
+
+func (e *Error) Unwrap() error {
+	if e.errno != 0 {
+		return syscall.Errno(e.errno)
+	}
+
+	return nil
+}
+
+// Backtrace flush backtrace of underlying error to string.
+//
+// Its not passed to realization of Error interface on purpose since
+// the main error should remain clear and simple
+func (err *Error) Backtrace() string {
+	buf := strings.Builder{}
+	buf.Reset()
+
+	for _, line := range err.backtrace {
+		if line.sName != "" {
+			buf.WriteString(fmt.Sprintf("'%s'@", line.sName))
+		}
+		buf.WriteString(fmt.Sprintf("<0x%x>+0x%x\n", line.sAddress, line.ip-line.sAddress))
+		if line.sFile != "" {
+			buf.WriteString(fmt.Sprintf("  in file '%s':%d\n", line.sFile, line.sLineno))
+		}
+	}
+
+	return buf.String()
+}
+
+func newError(e cErrorPtr) *Error {
+	realError := (*errorType)(e)
+
+	err := &Error{}
+	err.errno = uint64(realError.savedErrno)
+	if realError.description != nil {
+		err.description = C.GoString(realError.description)
+	}
+
+	if realError.backtrace != nil {
+		length := uintptr(realError.backtrace.length)
+		size := unsafe.Sizeof(backtraceEntry{})
+		for i := uintptr(0); i < length; i++ {
+			head := (*backtraceEntry)(unsafe.Pointer(uintptr(unsafe.Pointer(realError.backtrace.head)) + size*i))
+			line := backtraceLine{}
+			line.sAddress = head.symbolAddress
+			line.ip = head.ip
+			if head.symbolName != nil {
+				line.sName = C.GoString(head.symbolName)
+			}
+			if head.symbolFile != nil {
+				line.sFile = C.GoString(head.symbolFile)
+				line.sLineno = uint32(head.symbolLineno)
+			}
+
+			err.backtrace = append(err.backtrace, line)
+		}
+	}
+
+	return err
+}
+
+func handleErr(t pointerType, ptr unsafe.Pointer) (err error) {
+	ptrType := C.pathrs_type_t(t)
+	errPtr := cErrorPtr(C.pathrs_error(ptrType, ptr))
+	if errPtr != nil {
+		err = newError(errPtr)
+	}
+	free(ptrTypeError, errPtr)
+
+	return
+}
+
+func free(ptrType pointerType, ptr unsafe.Pointer) {
+	typeVal := C.pathrs_type_t(ptrType)
+	C.pathrs_free(typeVal, ptr)
+}
+
+type pointerType = int
+
+const (
+	ptrTypeNone   pointerType = C.PATHRS_NONE
+	ptrTypeError  pointerType = C.PATHRS_ERROR
+	ptrTypeRoot   pointerType = C.PATHRS_ROOT
+	ptrTypeHandle pointerType = C.PATHRS_HANDLE
+)
+
+type resolverType = uint64
+
+const (
+	kernelResolver   resolverType = C.PATHRS_KERNEL_RESOLVER
+	emulatedResolver resolverType = C.PATHRS_EMULATED_RESOLVER
+)
+
+type cRootPtr = unsafe.Pointer
+type cErrorPtr = unsafe.Pointer
+type cHandlePtr = unsafe.Pointer
+
+type fd = int
+
+type errorType struct {
```

Comment:
This question applies to all of these structs -- why not use the `C.pathrs_error_t` and similar cgo-generated types (in case you didn't know, cgo generates Go structs based on the C structs defined in `import "C"` source). I have a feeling quite a few of the `unsafe.Pointer` usages are actually incorrect because Go doesn't guarantee that their structs have identical layouts to C -- only the cgo ones are guaranteed to work.